### ISSUE TITLE
Finetuning on custom pick and place dataset on SO-101 arm

### ISSUE THREAD

**Issue by AkashKarnatak** [Robbyant/lingbot-va#29]
Hi, thank you for this project.

I attempted to fine-tune the base model (robbyant/lingbot-va-base) on my custom pick-and-place SO-101 arm dataset (https://huggingface.co/datasets/AkashKarnatak/pick-n-place-sq). I first converted the dataset into the required training format and was able to train the model successfully without any errors. The dataset contains 83 episodes, each approximately 6-7 seconds long. I trained the base model for 50k steps on this data.

However, when I deployed the trained model on the arm, the robot performed very random and incoherent movements. The generated rollout videos also show similarly random motion. This makes me suspect that the video model may be under-trained, but I am unsure whether 50k steps should already be sufficient for a dataset of this size. Below is what the generated video looks like.

https://github.com/user-attachments/assets/faa0d5be-7bd8-4db4-83b7-87cd820c73fa

I would appreciate some guidance on what might be going wrong. Could this behavior be due to insufficient training steps, or is it more likely an issue with data formatting, normalization, action scaling, or representation? Additionally, is a dataset of this size adequate for meaningful fine-tuning, or would significantly more data typically be required? Any debugging suggestions or recommended training configurations would be very helpful.

Thank you in advance for your help.

**Comment by zqh0253** [comment]
Thanks for sharing this! Could you provide the training and validation loss curves? It would also be helpful if you could try generating a video from both a training set image and a validation set image. Based on the current results, I suspect the model might be overfitting.

**Comment by AkashKarnatak** [comment]
<img width="1596" height="707" alt="Image" src="https://github.com/user-attachments/assets/edc7de34-7b14-46d1-bd61-5b5b0ee22a0a" />

The above screenshot shows my training loss curve. I don’t think I have a validation set because the training script uses the entire dataset during training.

I am currently running the model without computing intermediate KV caches as when I compute the KV cache, the robot makes a lot of weird and random movements. The above generated video was from my inference code that doesn't use kv cache and it had a bug as I was not advancing frame_st_id. After fixing that, it worked better. But still, the accuracy is quite low. It completed the task successfully only once, and that also seemed accidental.

I am attaching both the generated and real videos for reference. I want to understand why the model does not work properly when computing KV cache. Could you please give some suggestions?

https://github.com/user-attachments/assets/944cdb64-2061-47b2-9330-17faadb8bf13

https://github.com/user-attachments/assets/e17ee544-8d0f-400f-ade1-6391b55275e3

**Comment by zqh0253** [comment]
What do you mean by "without computing intermediate KV caches"? 

Also, could you try generating a video from a training set image?

**Comment by AkashKarnatak** [comment]
You know how when we deploy policy after every inference step we need to compute kv cache of newly obtained real observation. I don't do that. A little code snippet might help 

```py
# action loop
start = time.time()
while time.time() < start + 300: 
    print("Infering...")
    actions = infer(policy, obs)
    print("Executing...")
    real_obs = execute_action(actions)

    obs = make_observation(real_obs)
    # print('Computing KV cache...')
    # compute_kv_cache(policy, obs)
```

Generating video on training image also yields similar video.

https://github.com/user-attachments/assets/9b51a8ea-3039-4b42-9934-f74ca22a2799

**Comment by zqh0253** [comment]
Can you share your training config? The loss is higher than expected. Some config might be wrong.

**Comment by AkashKarnatak** [comment]
Below is the config I used for training. Also trained this on a single H100 GPU with batch size of 1.

```py
# Copyright 2024-2025 The Robbyant Team Authors. All rights reserved.
from easydict import EasyDict

from .shared_config import va_shared_cfg

va_robotwin_cfg = EasyDict(__name__='Config: VA robotwin')
va_robotwin_cfg.update(va_shared_cfg)

va_robotwin_cfg.wan22_pretrained_model_name_or_path = "robbyant/lingbot-va-base"

va_robotwin_cfg.attn_window = 72
va_robotwin_cfg.frame_chunk_size = 2
va_robotwin_cfg.env_type = 'robotwin_tshape'

va_robotwin_cfg.height = 256
va_robotwin_cfg.width = 320
va_robotwin_cfg.action_dim = 30
va_robotwin_cfg.action_per_frame = 24
va_robotwin_cfg.obs_cam_keys = [
    'observation.images.top', 'observation.images.wrist',
    'observation.images.right'
]
va_robotwin_cfg.guidance_scale = 5
va_robotwin_cfg.action_guidance_scale = 1

va_robotwin_cfg.num_inference_steps = 25
va_robotwin_cfg.video_exec_step = -1
va_robotwin_cfg.action_num_inference_steps = 50

va_robotwin_cfg.snr_shift = 5.0
va_robotwin_cfg.action_snr_shift = 1.0

va_robotwin_cfg.used_action_channel_ids = list(range(0, 6))
inverse_used_action_channel_ids = [
    len(va_robotwin_cfg.used_action_channel_ids)
] * va_robotwin_cfg.action_dim
for i, j in enumerate(va_robotwin_cfg.used_action_channel_ids):
    inverse_used_action_channel_ids[j] = i
va_robotwin_cfg.inverse_used_action_channel_ids = inverse_used_action_channel_ids

va_robotwin_cfg.action_norm_method = 'quantiles'
va_robotwin_cfg.norm_stat = {
    "q01": [
        -84.77347355027422,
        -98.66407012292476,
        -41.42723397782268,
        57.94535129798711,
        -15.675177242282363,
        1.0866095025389453
    ] + [0.] * 24,
    "q99": [
        10.620106315796306,
        22.04794919505999,
        92.59867317939553,
        99.16598951485992,
        4.917402314765924,
        99.99093274789122
    ] + [0.] * 24,
}
```

```py
# Copyright 2024-2025 The Robbyant Team Authors. All rights reserved.
from easydict import EasyDict
from .va_lerobot_cfg import va_robotwin_cfg
import os

va_robotwin_train_cfg = EasyDict(__name__='Config: VA robotwin train')
va_robotwin_train_cfg.update(va_robotwin_cfg)

# va_robotwin_train_cfg.resume_from = '/robby/share/Robotics/lilin1/code/Wan_VA_Release/train_out/checkpoints/checkpoint_step_10'

va_robotwin_train_cfg.dataset_path = '/home/ubuntu/workspace/lingbot-va/data/pick-n-place-sq'
va_robotwin_train_cfg.empty_emb_path = os.path.join(va_robotwin_train_cfg.dataset_path, 'empty_emb.pt')
va_robotwin_train_cfg.enable_wandb = True
va_robotwin_train_cfg.load_worker = 16
va_robotwin_train_cfg.save_interval = 1000
va_robotwin_train_cfg.gc_interval = 50
va_robotwin_train_cfg.cfg_prob = 0.1

# Training parameters
va_robotwin_train_cfg.learning_rate = 1e-5
va_robotwin_train_cfg.beta1 = 0.9
va_robotwin_train_cfg.beta2 = 0.95
va_robotwin_train_cfg.weight_decay = 0.1
va_robotwin_train_cfg.warmup_steps = 10
va_robotwin_train_cfg.batch_size = 1 
va_robotwin_train_cfg.gradient_accumulation_steps = 1
va_robotwin_train_cfg.num_steps = 50000 
```

**Comment by lilin-hitcrt** [comment]
Hi! Thanks for reporting this issue.

I've converted your dataset to the proper format and tested it with our training code. Following the provided data format and default parameters, the model trains successfully on 8 GPUs and shows good visualization results after ~200 steps.

Formatted dataset and visualization results (200 steps):
https://drive.google.com/file/d/1D52nK4ZOJmWBXKv1nWrLb9YBwq8nKa_b/view?usp=sharing

https://github.com/user-attachments/assets/c960f27c-400c-4d14-a850-e4735ab9f840

The training is working as expected. Feel free to continue training for better results. Let me know if you have any questions!

**Comment by lilin-hitcrt** [comment]
This is the loss curve:

<img width="2216" height="664" alt="Image" src="https://github.com/user-attachments/assets/42127e46-5385-4c22-85dc-93e581fc9134" />

**Comment by AkashKarnatak** [comment]
Hi @lilin-hitcrt 

I don't think there were any issues with my data conversion. I would like to know the training config parameters you used. Could you please share the training scripts and config files?

**Comment by lilin-hitcrt** [comment]
> Hi [@lilin-hitcrt](https://github.com/lilin-hitcrt)
> 
> I don't think there were any issues with my data conversion. I would like to know the training config parameters you used. Could you please share the training scripts and config files?

I have updated the code, please check wan_va/configs/va_demo_train_cfg.py for debug.

**Comment by AkashKarnatak** [comment]
Thank you. I'll continue the training and will try evaluating the model on real setup as well.

**Comment by AkashKarnatak** [comment]
Hello @lilin-hitcrt 

Thank you for sharing the configs. The training curve looks much better now. I trained the model for 1000 steps on 8xH100, and the loss curve is shown below.

<img width="1920" height="847" alt="Image" src="https://github.com/user-attachments/assets/157c9176-62c5-423e-912c-38d1c192cf42" />

The video generation quality is definitely better, but the action generation seems worse. I am still not able to get good closed-loop performance. By closed loop, I mean computing the KV cache on newly obtained real observations. By open loop, I mean skipping the KV cache computation.

Below are the results when I deployed the 1000-step checkpoint on the robot.

## Open loop
### Generated

https://github.com/user-attachments/assets/9c2123c4-4019-4c73-b18c-c7d8de5e6fad

### Real

https://github.com/user-attachments/assets/8a5139d0-9ef2-4f2b-a822-700294fefb38

## Closed loop
### Generated 

https://github.com/user-attachments/assets/040abaa3-4558-4581-b922-9a4d2154c421

### Real

https://github.com/user-attachments/assets/add8b7e7-a039-422c-9d10-c20228bda0c7

In the open-loop deployment, the robot makes random back-and-forth movements. The accuracy is quite low, but I was somehow able to record one successful trial.

In the closed-loop deployment, the behavior is very chaotic. It was very difficult to even record a single episode without crashing the camera.

Below is my inference script:

```py
import time
import torch
from einops import rearrange
from websocket_client import WebsocketClientPolicy
from lerobot.cameras.opencv.configuration_opencv import OpenCVCameraConfig
from lerobot.robots.so101_follower import SO101Follower, SO101FollowerConfig

prompt = "Pick the green cube and place it inside the blue box"

states = [
    "shoulder_pan.pos",
    "shoulder_lift.pos",
    "elbow_flex.pos",
    "wrist_flex.pos",
    "wrist_roll.pos",
    "gripper.pos",
]


def init_robot():
    robot_config = SO101FollowerConfig(
        port="/dev/ttyACM0",
        id="follower",
        cameras={
            "top": OpenCVCameraConfig(
                index_or_path=0,
                width=1080,
                height=1080,
                fps=30,
            ),
            "wrist": OpenCVCameraConfig(
                index_or_path=1, width=640, height=480, fps=30, fourcc="MJPG"
            ),
        },
    )

    robot = SO101Follower(robot_config)
    robot.connect()

    return robot


def reset_pos(robot):
    pos = {
        "shoulder_pan.pos": -5.569900405754339,
        "shoulder_lift.pos": -99.23501912452188,
        "elbow_flex.pos": 100.0,
        "wrist_flex.pos": 74.46623093681916,
        "wrist_roll.pos": 2.076716344979232,
        "gripper.pos": 95.98366235534377,
    }
    robot.send_action(pos)
    time.sleep(3)


def make_observation(robot, obs):
    global states, prompt

    if not obs:
        obs = robot.get_observation()
        observation = {
            "obs": [
                {
                    "observation.images.top": obs["top"],
                    "observation.images.wrist": obs["wrist"],
                }
            ],
            "state": torch.tensor([obs[state] for state in states])
            .view(-1, 1, 1)
            .expand(-1, 1, 8)
            .numpy(),
            "prompt": [prompt],
            "reset": False,
            "compute_kv_cache": False,
        }
    else:
        observation = {
            "obs": [
                {
                    "observation.images.top": ob["top"],
                    "observation.images.wrist": ob["wrist"],
                }
                for ob in obs[::-2] # 30fps -> 15fps
            ],
            "state": torch.tensor([[ob[state] for state in states] for ob in obs])
            .permute(0, 1)
            .view(6, -1, 8)
            .numpy(),
            "prompt": [prompt],
            "reset": False,
            "compute_kv_cache": False,
        }

    return observation


def execute_action(robot, actions):
    global states

    observations = []
    for action in actions:
        action = {s: a for s, a in zip(states, action)}
        robot.send_action(action)

        obs = robot.get_observation()
        observations.append(obs)

        time.sleep(1 / 30)

    return observations


def reset(policy, obs):
    obs["reset"] = True
    policy.infer(obs)


def infer(policy, obs):
    obs["reset"] = False
    obs["compute_kv_cache"] = False
    actions = policy.infer(obs)["action"]
    return rearrange(actions, "c t f -> (f t) c")


def compute_kv_cache(policy, obs):
    obs["reset"] = False
    obs["compute_kv_cache"] = True
    policy.infer(obs)


def main():
    robot = init_robot()
    reset_pos(robot)

    policy = WebsocketClientPolicy(port=29056)

    # reset and init policy
    obs = make_observation(robot, None)
    reset(policy, obs)

    # action loop
    start = time.time()
    while time.time() < start + 300:
        print("Infering...")
        actions = infer(policy, obs)

        print("Executing...")
        real_obs = execute_action(robot, actions)

        obs = make_observation(robot, real_obs)
        print("Computing KV cache...")
        compute_kv_cache(policy, obs)

    robot.disconnect()


if __name__ == "__main__":
    main()
````

What do you think could be the issue with closed loop deployement?

**Comment by lilin-hitcrt** [comment]
1. Maybe “for ob in obs[::-2] # 30fps -> 15fps”  should be "for ob in obs[::2] # 30fps -> 15fps"?  
2. Make sure actions are normalized before input the model and denormalized when execute. 
3. The model appears to be overfitting at 1000 steps. 300-500 steps seems more appropriate. You can set aside some validation data during training.
4. Ensure you select the correct keyframes based on your video frame sampling rate.

**Comment by lilin-hitcrt** [comment]
For the open-loop experiment, you should update the KV cache with the generated latents rather than simply skipping the KV cache.

**Comment by AkashKarnatak** [comment]
> 1. Maybe “for ob in obs[::-2] # 30fps -> 15fps”  should be "for ob in obs[::2] # 30fps -> 15fps"?
> 
>     2. Make sure actions are normalized before input the model and denormalized when execute.
> 
>     3. The model appears to be overfitting at 1000 steps. 300-500 steps seems more appropriate. You can set aside some validation data during training.
> 
>     4. Ensure you select the correct keyframes based on your video frame sampling rate.

1) Thank you for pointing that out. I forgot to reverse the list. I wanted to sample frames at 15fps which includes the last frame, something like obs[::-2][::-1].
2) wan_va_server.py already has action preprocess and postprocess functions that normalizes using q01 and q99 so I don't think there is any issue with normalization.
3) I tested 600 steps checkpoint as well but it wasn't any better. Btw, how do you know it is overfitting from the loss curve?


> For the open-loop experiment, you should update the KV cache with the generated latents rather than simply skipping the KV cache.

In open loop I never clear cache pred cache so I guess it is fine.

**Comment by kingdom-of-warriors** [comment]
Hi there! Overall, your inference code looks mostly correct, but after cross-referencing it with the inference script provided in [issue #30](https://github.com/Robbyant/lingbot-va/issues/30), I noticed a few critical issues:

**The primary cause of the SO101 jittering is this line**: `return rearrange(actions, "c t f -> (f t) c")`. This completely scrambles the chronological order of the action execution sequence, which leads to the robotic arm shaking violently. You need to prioritize the temporal dimension (the chunk t must come before the frame f). It should be corrected to: `return rearrange(actions, "c t f -> (t f) c")`.

**Incorrect state input**: The state input fed to the model's KV Cache update should be the raw action tensor it predicted in the previous step, rather than the physical state read from the observation.

**Skipping the initial chunk**: The very first action chunk generated at the beginning of the inference loop should be discarded.

I've made some modifications based on your original script, and this updated version successfully executes the task on my end:
```python
import time
from einops import rearrange
from deploy.websocket_client_policy import WebsocketClientPolicy
import rclpy
import threading

from utils import ARXController

prompt = "Swap three distinct mugs under a coffee machine without tipping or repeating."

def make_observation(controller: ARXController, obs):
    global prompt

    if not obs:
        obs = controller.get_observation()
        observation = {
            "obs": [
                {
                    "observation.images.head": obs["image"],
                }
            ],
            "prompt": [prompt],
            "reset": False,
            "compute_kv_cache": False,
        }
    else:
        observation = {
            "obs": [
                {
                    "observation.images.head": ob["image"],
                }
                for ob in obs[::-2][::-1]
            ],
            "prompt": [prompt],
            "reset": False,
            "compute_kv_cache": False,
        }

    return observation


def execute_action(controller: ARXController, actions):
    observations = []
    for action in actions:
        # import ipdb; ipdb.set_trace()
        controller.execute_action(action)
        obs = controller.get_observation()
        observations.append(obs)
        time.sleep(1 / 30) 

    return observations


def reset(policy, obs):
    obs["reset"] = True
    obs["n_view"] = 1
    policy.infer(obs)

def infer(policy, obs):
    obs["reset"] = False
    obs["compute_kv_cache"] = False
    ret = policy.infer(obs)
    raw_action = ret["action"] # [C, T, F] -> [7, 4, 8]
    parsed_actions = rearrange(raw_action, "c t f -> (t f) c")
    
    return parsed_actions, raw_action


def compute_kv_cache(policy, obs, raw_action):
    obs["reset"] = False
    obs["compute_kv_cache"] = True
    obs["imagine"] = False
    obs["state"] = raw_action
    policy.infer(obs)


def main():
    rclpy.init()
    controller = ARXController()
    
    spin_thread = threading.Thread(target=rclpy.spin, args=(controller,), daemon=True)
    spin_thread.start()
    time.sleep(2)
    print("✓ ROS initialized")
    
    controller.reset()
    policy = WebsocketClientPolicy(host="127.0.0.1", port=29536)
    
    current_obs_dict = make_observation(controller, None)
    reset(policy, current_obs_dict)

    start = time.time()
    first = True
    
    while time.time() < start + 600:
        print("Infering...")
        actions_to_execute, raw_action_tensor = infer(policy, current_obs_dict)
        
        if first:
            f_dim = raw_action_tensor.shape[2]
            actions_to_execute = actions_to_execute[f_dim:]
        
        print(f"Executing {len(actions_to_execute)} steps...")
        real_obs_history = execute_action(controller, actions_to_execute)

        print("Computing KV cache...")
        kv_obs_dict = make_observation(controller, real_obs_history)
        
        compute_kv_cache(policy, kv_obs_dict, raw_action_tensor)
        current_obs_dict = make_observation(controller, None)
        
        first = False


if __name__ == "__main__":
    main()
```
